### ISSUE TITLE
federation: node-fetch in @apollo/gateway don't support unix: urls

### ISSUE THREAD

**Issue by lennyburdette** [apollographql/apollo-server#2835]
**Version:** `@apollo/gateway`@0.6.6

**Description:**
We use an envoy sidecar process to manage service-to-service routing. All requests go to the same unix socket and we add headers to tell envoy which service cluster to route to.

```js
const gateway = new ApolloGateway({
  serviceList: [
    { name: 'service-a', url: `unix:/path/to/envoy.sock:/graphql` },
    { name: 'service-b', url: `unix:/path/to/envoy.sock:/graphql` },
  ],
  buildService({ name, url }) {
    return new RemoteGraphQLDataSource({
      url,
      willSendRequest({ request }: any) {
        request.http.headers.set('Host', `${name}.specialenvoyheader`);
      }
    });
  }
});
```

Currently, this cannot work because [`getServiceDefinitionsFromRemoteEndpoint`](https://github.com/apollographql/apollo-server/blob/9ca984925820a56b6a4abf24a2134893ac816f65/packages/apollo-gateway/src/loadServicesFromRemoteEndpoint.ts#L29) in `@apollo/gateway` uses node-fetch, which throws this error: `Only HTTP(S) protocols are supported`.

We use unix sockets because we can chown the socket and restrict access to it, unlike a tcp port. If we used TCP we'd have to also use TLS, which is overhead and complexity we'd like to avoid.

Adding support for unix sockets in node-fetch is marked as ["wontfix"](https://github.com/bitinn/node-fetch/issues/336).

I would suggest the [got](https://www.npmjs.com/package/got) library, which has the same API, but it has 58 transitive dependencies compared to node-fetch's zero. 😬

**Comment by lennyburdette** [comment]
I might be able to switch to TLS over TCP, but I'll need to be able to pass [`agent`](https://github.com/bitinn/node-fetch#custom-agent) to node-fetch so I can use our self-signed certs.

**Comment by jbaxleyiii** [comment]
Hey @lennyburdette! We are working on a way to provide the service definitions files instead of having to the do the fetch at startup. That plus providing a custom datasource should be enough for y'all yes?

**Comment by lennyburdette** [comment]
@jbaxleyiii That would probably do it! I'm blocked on building out a proof-of-concept for now, but it's good to know there's an alternative in the works.

If fetching the `_service.sdl` field lives on for certain use cases, it might be worth adding an `agent: http.Agent` option to the ApolloGateway constructor.
